### ISSUE TITLE
stream: function() support doesn't work as documented.

### ISSUE THREAD

**Issue by lagnat** [sidorares/node-mysql2#980]
The docs say that the function is async and should call a callback upon creation of the stream.

```
  stream: function (cb) {
    const newStream = new SocksConnection({host: 'remote.host', port: 3306}, {host: 'localhost', port: 1080});
    cb(null, newStream);
  }
```

  The code says the function is synchronous and should return the stream.

```
      // if stream is a function, treat it as "stream agent / factory"
      if (typeof opts.config.stream == 'function') {
        this.stream = opts.config.stream(opts);
      } else {
```

**Comment by sidorares** [comment]
Yeah, I think docs are wrong here, thanks for heads up! I'll try to fix

**Comment by perry-mitchell** [comment]
Any way we could have this updated? Having it synchronous means that we can't create stream-connections lazily, which would be preferred. This requires that the stream exists upon connection request from a connection pool.

**Comment by sidorares** [comment]
you can still do it lazily - return first "half" of the stream synchronously and connect events later when you have downstream connection, though it's not the best experience

**Comment by perry-mitchell** [comment]
@sidorares So you mean creating a duplex, returning that, and then connecting the new stream once its created asynchronously?

For now I've opted to just delaying the pool creation until I have 1 stream, and then just provide that. Ideally though this would support async creation so the passed function just acts as a factory.

**Comment by sidorares** [comment]
>  So you mean creating a duplex, returning that, and then connecting the new stream once its created asynchronously?

yes, something like that. I'm also happy to try to add support for `Promise<Stream>` result

**Comment by martenjacobs** [comment]
@sidorares I'm trying to understand your idea but it's not quite clear. Do you have any example code?
